machineconfiguration/v1alpha1/types_machineconfignode.go

```diff
@@ -0,0 +1,140 @@
+// Compatibility level 4: No compatibility is provided, the API can change at any point for any reason. These capabilities should not be used by applications needing long term support.
+// +openshift:compatibility-gen:level=4
+type MachineConfigNode struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// spec describes the configuration of this specific machineconfignode
+	// +kubebuilder:validation:Required
+	Spec MachineConfigNodeSpec `json:"spec"`
+
+	// status describes the last observed state of this machineconfignode
+	// +kubebuilder:validation:Optional
+	// +optional
+	Status MachineConfigNodeStatus `json:"status"`
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// MachineConfigNodeList describes all of the MachinesStates on the system
+//
+// Compatibility level 4: No compatibility is provided, the API can change at any point for any reason. These capabilities should not be used by applications needing long term support.
+// +openshift:compatibility-gen:level=4
+type MachineConfigNodeList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata"`
+
+	Items []MachineConfigNode `json:"items"`
+}
+
+// MCOObjectReference holds information about an object the MCO either owns
+// or modifies in some way
+type MCOObjectReference struct {
+	// kind describes the type of object thsi reference points to
+	// +kubebuilder:validation:Required
+	// +required
+	Kind string `json:"kind"`
+	// name describes what this object is called
+	// +kubebuilder:validation:Required
+	// +required
+	Name string `json:"name"`
+	// apiVersion details the group and version of this resource.
+	// +kubebuilder:validation:Required
+	// +required
+	APIVersion string `json:"apiVersion"`
+}
+
+// MachineConfigNodeSpec describes the type of State we are managing
+type MachineConfigNodeSpec struct {
+	// nodeRef contains references to the node for this MachineConfigNode
+	// +optional
+	NodeRef MCOObjectReference `json:"nodeRef"`
+
+	// pool is the MachineConfigPool this MachineConfigNode is describing
+	// +optional
+	Pool MCOObjectReference `json:"pool"`
+}
+
+// MachineConfigNodeStatus holds the reported information on a particular MachineConfigNode
+type MachineConfigNodeStatus struct {
+	// conditions represent the observations of a MachineConfigNode's current state.
+	// Known .status.conditions.type are: "UpdatePreparing", "UpdateInProgress", "UpdatePostAction", "UpdateCompleting", "Resuming", "Updated",
+	// "Ready", "Errored", "ComparingMCs", "DrainingNode", "ApplyingFilesAndOS", "CordoningNode", "RebootingNode", and "ReloadingCRIO"
+	// +patchMergeKey=type
+	// +patchStrategy=merge
+	// +listType=map
+	// +listMapKey=type
+	Conditions []metav1.Condition `json:"conditions,omitempty" patchStrategy:"merge" patchMergeKey:"type"`
+	// observedGeneration represents the generation observed by the controller.
```

Comment:
moving desired and current cfg to the spec so I will keep this. And yes your assumption is right. Once the current/desired config change this will be updated. (i have not implemented this part yet in my draft PR)

Reply:
See the other comment I just left, you want desired in spec, but then also desired and current in status. Current should not be in spec

Reply:
I think we need to expand this godoc to explain when and where this observed generation is coming from. When does this get updated, what triggers it, is it the observed generation of the MCN spec? None of this is obvious right now.

@wking Has some comments about this field on the EP we should try to resolve as well